### PR TITLE
Add Single Product Details block

### LINKED ISSUE #8186: Create the Single Product Details block
Create a block version of the product details (visible on the single product template), so merchants can add the product description, information, and reviews to their stores.

| Property | Value |
|---------|---------|
| Block title | Product Details |
| Block description | TBD |
| Block slug | `product-details` |
| Block file location | `assets/js/blocks/product-details/` |
| Block class location | `src/BlockTypes/ProductDetails.php` |
| Style | <ul> <li> Typography <li> Color <li> Spacing </ul> |
| Visibility | Woocommerce Core |

# Tabs

## Details:

| Property | Value |
|---------|---------|
| Equivalent template file | [`woocommerce/templates/single-product/tabs/tabs.php`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/single-product/tabs/tabs.php) |
| Equivalent template function | [`woocommerce_default_product_tabs()`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-template-functions.php#L1852) |

## Screenshot:

Screenshot 2023-01-15 at 20 19 39

# Description

## Details:

| Property | Value |
|---------|---------|
| Equivalent template file | [`woocommerce/templates/single-product/tabs/description.php`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/single-product/tabs/description.php) |
| Equivalent template function | [`the_content()`](https://github.com/WordPress/WordPress/blob/master/wp-includes/post-template.php#L245) |

## Screenshot:

Screenshot 2023-01-15 at 19 34 39

# Additional information

## Details:

| Property | Value |
|---------|---------|
| Equivalent template file | [`woocommerce/templates/single-product/tabs/additional-information.php`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/single-product/tabs/additional-information.php) |
| Equivalent template function | [`wc_display_product_attributes()`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-template-functions.php#L3462) |

## Screenshot:

Screenshot 2023-01-15 at 19 45 11

# Reviews

## Details:

| Property | Value |
|---------|---------|
| Equivalent template file | [`woocommerce/templates/single-product-reviews.php`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/single-product-reviews.php) |
| Equivalent template function (for listing existing reviews) | [`wp_list_comments()`](https://github.com/WordPress/WordPress/blob/master/wp-includes/comment-template.php#L2066) and [`woocommerce_comments()`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/wc-template-functions.php#L1936) |
| Equivalent template function (for submitting a new review) | [`comment_form()`](https://github.com/WordPress/WordPress/blob/master/wp-includes/comment-template.php#L2325) |

## Screenshot:

Screenshot 2023-01-15 at 19 45 29

### PR DESCRIPTION
**Pull request by thealexandrelara**
Create a block version of the product details (visible on the single product template), so merchants can add the product description, information, and reviews to their stores. In this first iteration, we are focusing in porting the existing functionality without allowing users to customize the block

Fixes woocommerce/woocommerce-blocks#8186

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

| Before | After |
| ------ | ----- |
|   image  |  image |

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [x] E2E tests

#### User Facing Testing

1. Enable a blocked theme in your WordPress installation (go to Appearance > Themes, and select one of the blocked themes options, such as: Twenty-Twenty Three);
2. Once the theme is enabled, go to Appearance > Editor (Beta);
3. On the left side menu, click on Templates;
4. Select the Single Product template, and click Edit;
5. Select the WooCommerce Single Product Block that comes by default with the template, remove it, and Save;
6. Using the Block Inserter, type: Single Product Details. Add it to the Editor and position the block as you want.
7. Make sure the Single Product Details block is being displayed and contains the tabs: Description, Additional Information and Reviews;
8. Save the changes. Access your website and click on a product;
9. Make sure the Single Product Details block appears and contains the product description (Description tab), product attributes (Additional Information tab) and the product reviews (Reviews tab).

### Changelog

> Add Single Product Details block that displays the product description, information, and reviews

### CODE REVIEW

File: assets/js/blocks/single-product-details/block.json
Comment (gigitux) [inline assets/js/blocks/single-product-details/block.json]: Should we name it `Product Details` or `Single Product Details`? 🤔  cc @vivialice

File: assets/js/blocks/single-product-details/block.json
Comment (gigitux) [inline assets/js/blocks/single-product-details/block.json]: Miss the declaration for the icon. Any idea about what icon should we pick? @vivialice

File: assets/js/blocks/single-product-details/block.tsx
Comment (gigitux) [inline assets/js/blocks/single-product-details/block.tsx]: I think that we should avoid using custom classes for the main wrapper. 

Also, it is better to use the `useBlockProps` hook.

```js
const blockProps = useBlockProps();
<div { ...blockProps }>
...
```

The class will be generated by GB and will be: `wp-block-${name-block}`. In this case, is `wp-block-woocommerce-single-product-details`.

File: assets/js/blocks/single-product-details/edit.tsx
Comment (gigitux) [inline assets/js/blocks/single-product-details/edit.tsx]: I think that you forgot to remove this console.log

File: assets/js/blocks/single-product-details/edit.tsx
Comment (gigitux) [inline assets/js/blocks/single-product-details/edit.tsx]: same

File: assets/js/blocks/single-product-details/edit.tsx
Comment (gigitux) [inline assets/js/blocks/single-product-details/edit.tsx]: For the same idea about "no custom class", we could refactor this code in this way:

```suggestion
		className: className ,
```

File: assets/js/blocks/single-product-details/index.tsx
Comment (gigitux) [inline assets/js/blocks/single-product-details/index.tsx]: It is not necessary to specify the attributes object again.

File: assets/js/blocks/single-product-details/style.scss
Comment (gigitux) [inline assets/js/blocks/single-product-details/style.scss]: If we resolve #8314, is this CSS still necessary?

File: bin/webpack-entries.js
Comment (gigitux) [inline bin/webpack-entries.js]: I think that it is not necessary. We want to ship this block.

File: src/BlockTemplatesController.php
Comment (gigitux) [inline src/BlockTemplatesController.php]: Is it necessary? 🤔

File: src/BlockTypes/SingleProductDetails.php
Comment (gigitux) [inline src/BlockTypes/SingleProductDetails.php]: This method is useful if Global Styles are applied, but this block doesn't support them. It is fine remove this.

Furthermore, we should add this code and attach the class name to the main wrapper

```php
		$classname          = $attributes['className'] ?? '';
```
The user can defined the custom classes here:
image

**Review by gigitux — CHANGES_REQUESTED** [review]
Thanks for working on this! Great job! I left some comments 💪

File: assets/js/blocks/single-product-details/index.tsx
Comment (gigitux) [inline assets/js/blocks/single-product-details/index.tsx]: We should make the block available only on the Single Product Template. I created a [util function](https://github.com/woocommerce/woocommerce-blocks/blob/e526eff91a4bd6e955e62793f5b18c259305e5ce/assets/js/atomic/utils/register-block-single-product-template.ts) to do this.

File: src/BlockTypesController.php
Comment (nefeline) [inline src/BlockTypesController.php]: How about disabling this block in Post and Page editor by not registering them, as we did for all other blocks [recently introduced](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/BlockTypesController.php#L253-L257)?

File: src/BlockTemplatesController.php
Comment (nefeline) [inline src/BlockTemplatesController.php]: This can be ditched

**Review by nefeline — CHANGES_REQUESTED** [review]
Hi there @thealexandrelara! Good to see this block shaping up here, well done! It works like a charm, but there's one caveat, the star ratings are still duplicated:

Screenshot 2023-01-31 at 20 39 50

Did you have the chance to look into this? I'm glad to jump on a call & try to figure this out if that helps!

Apart from that I'd also recommend addressing the PHP CS warnings.

File: src/BlockTemplatesController.php
Comment (thealexandrelara) [reply src/BlockTemplatesController.php]: You're correct, this is not necessary. I forgot to remove this piece of code that was put during my code investigation 😞

File: assets/js/blocks/single-product-details/edit.tsx
Comment (thealexandrelara) [reply assets/js/blocks/single-product-details/edit.tsx]: You're correct, fixed!

File: assets/js/blocks/single-product-details/block.tsx
Comment (thealexandrelara) [reply assets/js/blocks/single-product-details/block.tsx]: Ohh right, just fixed it

File: assets/js/blocks/single-product-details/edit.tsx
Comment (thealexandrelara) [reply assets/js/blocks/single-product-details/edit.tsx]: Solved!

File: bin/webpack-entries.js
Comment (thealexandrelara) [reply bin/webpack-entries.js]: Great! I was waiting for the confirmation if we would make any design changes before shipping this block, but I'll remove this flag 🙌. Thank you!

File: assets/js/blocks/single-product-details/index.tsx
Comment (thealexandrelara) [reply assets/js/blocks/single-product-details/index.tsx]: I think that function is not merged into trunk yet, right?

**Comment by thealexandrelara** [comment]
> Thanks for working on this! Great job! I left some comments 💪

Thank you for reviewing it! I'll be addressing all of those and let's wait for @vivialice to decide about the ones related to UX.

> Hi there @thealexandrelara! Good to see this block shaping up here, well done! It works like a charm, but there's one caveat, the star ratings are still duplicated:
> 
> Screenshot 2023-01-31 at 20 39 50
> 
> Did you have the chance to look into this? I'm glad to jump on a call & try to figure this out if that helps!
> 
> Apart from that I'd also recommend addressing the PHP CS warnings.

Hey @nefeline, thank you for your review! Yeah, I took a look into this, this issue occurs if you put the block in the Single Product template together with the classic template for the WooCommerce Single Product Block. To avoid that, I suggest to remove the default Single Product Block from the Single Product template before testing (step 5 of the User Facing Testing section in this PR)

About the PHP CS warnings, I'm going to address them now, thank you for pointing this out!

File: assets/js/blocks/single-product-details/index.tsx
Comment (thealexandrelara) [reply assets/js/blocks/single-product-details/index.tsx]: Nice, removed it!

File: src/BlockTypes/SingleProductDetails.php
Comment (thealexandrelara) [reply src/BlockTypes/SingleProductDetails.php]: Ohh good to know, thank you for sharing!

File: src/BlockTypesController.php
Comment (thealexandrelara) [reply src/BlockTypesController.php]: This is smart! Thank you for sharing, I just fixed it

File: src/BlockTemplatesController.php
Comment (thealexandrelara) [reply src/BlockTemplatesController.php]: I removed this piece of code! Thank you!

File: assets/js/blocks/single-product-details/style.scss
Comment (thealexandrelara) [reply assets/js/blocks/single-product-details/style.scss]: I hope we don't need it anymore after solving that issue; however, if we need to change some things in the UX I think we still are going to need this

**Comment by nefeline** [comment]
> Hey @nefeline, thank you for your review! Yeah, I took a look into this, this issue occurs if you put the block in the Single Product template together with the classic template for the WooCommerce Single Product Block.

Do you think it would be a good idea to ensure this block renders as expected in this scenario too (with the block rendering alongside the original template)? I ask because:

1. Nothing prevents store owners from doing so and having both in place.
2. I can imagine this is a possible scenario, even if it is just while the store owner is setting up their single product template with the new blocks and they wanna compare both layouts side by side before removing the original template altogether.

**Comment by thealexandrelara** [comment]
> > Hey @nefeline, thank you for your review! Yeah, I took a look into this, this issue occurs if you put the block in the Single Product template together with the classic template for the WooCommerce Single Product Block.
> 
> Do you think it would be a good idea to ensure this block renders as expected in this scenario too (with the block rendering alongside the original template)? I ask because:
> 
> 1. Nothing prevents store owners from doing so and having both in place.
> 2. I can imagine this is a possible scenario, even if it is just while the store owner is setting up their single product template with the new blocks and they wanna compare both layouts side by side before removing the original template altogether.

I'm happy to investigate what is occurring and check if it is a quick fix. I think we should time-box this because we should focus on blockfying the single product template for this first iteration. We need to consider that the design could change, including the Reviews tab part, since we already have a Reviews by Product block; maybe to keep consistency, we should replace the one we are currently using in this PR with that block, so if it takes a lot of time and effort to solve, the option to replace with the block we already have would work well (at least I hope so 😅 )

**Comment by nefeline** [comment]
> I'm happy to investigate what is occurring and check if it is a quick fix. I think we should time-box this because we should focus on blockfying the single product template for this first iteration. We need to consider that the design could change, including the Reviews tab part, since we already have a Reviews by Product block; maybe to keep consistency, we should replace the one we are currently using in this PR with that block, so if it takes a lot of time and effort to solve, the option to replace with the block we already have would work well (at least I hope so 😅 )

Cool! If you need a second pair of eyes here, say the word ;) ! We could also open an issue and solve it separately, but if we go this route, I suggest checking/confirming with the team if this issue is a blocker for releasing this block (just to be safe).

**Comment by thealexandrelara** [comment]
> Cool! If you need a second pair of eyes here, say the word ;) ! We could also open an issue and solve it separately, but if we go this route, I suggest checking/confirming with the team if this issue is a blocker for releasing this block (just to be safe).

I was investigating the issue and I found that it seems to be caused by the selector `id` [in the `single-product-reviews.php` file in WooCommerce](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/templates/single-product-reviews.php#L124). The id for the select element is set to 'rating':
```
<select name="rating" id="rating" required>
```

If we change that id to another thing, the issue is gone, like this:
```
<select name="rating" id="rating-selector" required>
```

It seems we are also using [the same id to another select element](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/admin/meta-boxes/class-wc-meta-box-product-reviews.php#L27), and it is probably causing a conflict which makes the ratings stars to appear duplicated. Do you know if we can use some kind of hook to manipulate this behavior and change the `id` for that select element?

File: assets/js/blocks/single-product-details/block.json
Comment (nerrad) [reply assets/js/blocks/single-product-details/block.json]: Just name it `Product Details` for now.

**Comment by gigitux** [comment]
I think that addressing the issue could require time, and I don't believe that it makes sense to make this effort (at least now), especially because:
- This block is an alternative to the Classic Single Product Block. Use the block with the Classic Product Block is an edge case.
- In the next future, we will deprecate the Classic Product Block.

Any thoughts? @nerrad @tjcafferkey @danielwrobert

**Comment by gigitux** [comment]
Should we put this block inside the folder `product-elements`? Furthermore, it could make sense to rename the block name (and relative folder) into `product-details`. What do you think? @nefeline

**Comment by tjcafferkey** [comment]
>I think that addressing the issue could require time, and I don't believe that it makes sense to make this effort (at least now), especially because:
> * This block is an alternative to the Classic Single Product Block. Use the block with the Classic Product Block is an edge case.
> * In the next future, we will deprecate the Classic Product Block.

I think if this requires a significant time investment then it's fine to ship. We can later work out whether it's worth fixing before deprecating the Classic Product block depending on at what point in the future that will be happening.

But yes, it seems that this issue only occurs in a specific scenario which I would consider an edge case.

**Comment by nefeline** [comment]
> Should we put this block inside the folder product-elements?

By product elements, do you mean within `woocommerce-blocks/assets/js/atomic/blocks/product-elements` ? That's a good question; I'm not sure: does this block have all prerequisites to be classified as atomic? If we follow this route, we might have to move the Add to Cart Form block woocommerce/woocommerce-blocks#8284 and any other blocks that belong to the single product template to this directory as well for consistency.

> Furthermore, it could make sense to rename the block name (and relative folder) into product-details. What do you think? @nefeline

Naming it Product Details and updating the relative folder makes sense to me 👍 .

Regarding the duplicate star ratings issue:

> I think if this requires a significant time investment then it's fine to ship. We can later work out whether it's worth fixing before deprecating the Classic Product block depending on at what point in the future that will be happening.
> But yes, it seems that this issue only occurs in a specific scenario which I would consider an edge case.

I have created woocommerce/woocommerce#42475, so we can keep track: I'm not entirely sure if this is something that takes a significant amount of time to fix as I didn't investigate it myself; there could be alternative paths/workarounds to explore. I'll spend some time on it between today/tomorrow just in case: if no quick fix is available, then I agree it makes sense to work on it later as it occurs only under specific circumstances.

**Comment by gigitux** [comment]
> By product elements, do you mean within `woocommerce-blocks/assets/js/atomic/blocks/product-elements` ? That's a good question; I'm not sure: does this block have all prerequisites to be classified as atomic? If we follow this route, we might have to move the Add to Cart Form block woocommerce/woocommerce-blocks#8284 and any other blocks that belong to the single product template to this directory as well for consistency.

Technically, there isn't any difference between atomic blocks and other blocks. For now, I would put all of them into the `product-elements`. The reason is that the blocks used by the Single Product Block (and now by the Products block) are in this folder. We could always revisit this choice. 

Via [96fab80](https://github.com/woocommerce/woocommerce-blocks/pull/8225/commits/96fab8035cb9993b92dd982164b4b6eb3b5b4d40), I addressed those two points. Furthermore,  I created a follow-up about icon woocommerce/woocommerce-blocks#8427.


@nefeline, could you do another look?

File: src/BlockTypes/ProductDetails.php
Comment (nefeline) [inline src/BlockTypes/ProductDetails.php]: The styles are now broken due to the latest changes to rename the block from `single-product-details` to `product-details`. The following should do the trick:
```suggestion
			'<div class="wc-block-woocommerce-product-details %1$s">
```
I also suggest renaming the class prefix from `wp-block` to `wc-block` so it is aligned with the standards we are using in other blocks.

File: assets/js/atomic/blocks/product-elements/product-details/style.scss
Comment (nefeline) [inline assets/js/atomic/blocks/product-elements/product-details/style.scss:1]:
```suggestion
.wc-block-woocommerce-product-details {
```

**Review by nefeline — CHANGES_REQUESTED** [review]

**Comment by nefeline** [comment]
Nice work! Apart from the issue with the styles flagged in the review, everything else looks a-ok 👍 

> Technically, there isn't any difference between atomic blocks and other blocks. 

Isn't it the case that the atomic blocks for Product Elements are supposed to be used as children of contextual blocks? In other words, the context can be used to [pass product data down to all children blocks in a given tree](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/shared/context/product-data-context.tsx)?

**Comment by nefeline** [comment]
I just noticed some oddities: there are two console errors related to the `product-details` block and the `product-image-gallery`:

Screenshot 2023-02-15 at 09 24 03

In the context of this block (the Product Details), I suspect it might be related to the recent changes that migrated this block to `woocommerce-blocks/assets/js/atomic/blocks/product-elements` as it didn't occur previously (the error is only visible after [96fab80](https://github.com/woocommerce/woocommerce-blocks/pull/8225/commits/96fab8035cb9993b92dd982164b4b6eb3b5b4d40))

**Comment by gigitux** [comment]
> Nice work! Apart from the issue with the styles flagged in the review, everything else looks a-ok 👍
> 
> > Technically, there isn't any difference between atomic blocks and other blocks.
> 
> Isn't it the case that the atomic blocks for Product Elements are supposed to be used as children of contextual blocks? In other words, the context can be used to [pass product data down to all children blocks in a given tree](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/shared/context/product-data-context.tsx)?

Mmm, good observation. All the more reason we have to add the blocks in the product-elements block: the Single Product block will use those blocks :D 

> I just noticed some oddities: there are two console errors related to the product-details block and the product-image-gallery:


Fixed for the Single Product Details, I will create a dedicated issue for the Product Image Gallery woocommerce/woocommerce-blocks#8436! Good catch!

EDIT: My first attempt was [43dd558](https://github.com/woocommerce/woocommerce-blocks/pull/8225/commits/43dd55847835fe016fe4d5e950baa5d4bd5bc2d5), but after worked on woocommerce/woocommerce-blocks#8445 I realized that there is a better approach. Check [85e6247](https://github.com/woocommerce/woocommerce-blocks/pull/8225/commits/85e6247e04588a4929e1d2d32d3bb6cefd5c7fa7).

File: src/BlockTypes/ProductDetails.php
Comment (gigitux) [reply src/BlockTypes/ProductDetails.php]: We decided to avoid adding a custom class. We should use WP generated class (more context: p1675160559571929-slack-C02UBB1EPEF)

Comment (nefeline) [reply src/BlockTypes/ProductDetails.php]: > We decided to avoid adding a custom class. We should use WP generated class (more context: p1675160559571929-slack-C02UBB1EPEF)

Oh cool, I missed that convo, thanks for sharing! I'll also go ahead and update #8284 with this prefix.

In case someone else comes across this convo: the reason for this decision was that for new blocks, it's a good idea not to introduce new classes, as block themes interact with blocks via theme.json.

File: assets/js/atomic/blocks/product-elements/product-details/style.scss
Comment (nefeline) [reply assets/js/atomic/blocks/product-elements/product-details/style.scss:1]: see https://github.com/woocommerce/woocommerce-blocks/pull/8225#discussion_r1106985254 for context.

**Review by nefeline — APPROVED** [review]
Thanks for addressing all the feedback @gigitux ! Confirmed that:

- Styles are fixed.
- The console 404 error is fixed.

Ref. this discussion:

> Mmm, good observation. All the more reason we have to add the blocks in the product-elements block: the Single Product block will use those blocks :D

Considering that you already moved this block and also the gallery block to this directory, sounds like it makes sense to have the Add to cart form block moved as well, right? I'll open a separate PR to work on this so we can unblock & merge  #8284

**Comment by gigitux** [comment]
Thanks for your great review! (and thanks @thealexandrelara for working on this block).

> Considering that you already moved this block and also the gallery block to this directory, sounds like it makes sense to have the Add to cart form button moved as well, right? I'll open a separate PR to work on this so we can unblock & merge woocommerce/woocommerce-blocks#8284

It makes sense! 👍

**Pull request merged**
